File: Questions-Solutions/Two-Dimensional-Arrays/Q02MineSweeper.cpp
#include <iostream>
#include <vector>
#include <queue>

std::vector<std::vector<int>> click(std::vector<std::vector<int>> field,int i,int j){
  std::queue<std::vector<int>> toCheck;

  if(field[i][j]==0){

    field[i][j] = -2;
    toCheck.push({i,j});
  }
  std::vector<int> pair;
  while(!toCheck.empty()){
    pair=toCheck.front();
    i= pair[0];
    j=pair[1];
    toCheck.pop();
    for(int a=i-1;a<=i+1;a++){
      for(int b=j-1;b<=j+1;b++){
        if(a>=0 && b>=0 && a<5 && b<5 && field[a][b]==0){
          field[a][b]=-2;
        }
      }
    }
  }
  return field;
}


void mineSweeper(int bombs[3][2],int num_cols,int num_rows){
  std::vector<std::vector<int>> field(5,std::vector<int>(5));

  int numOfBombs = 3;


  for(int i=0;i<numOfBombs;i++){
    field[bombs[i][0]][bombs[i][1]]=-1;
    for(int a=(bombs[i][0]-1);a<=(bombs[i][0]+1);a++){
      for(int b=(bombs[i][1]-1);b<=(bombs[i][1]+1);b++){
        if(a>=0 && b>=0 && a<num_rows && b<num_cols && field[a][b]!=-1){
          field[a][b]++;
        }
      }
    }
  }

  field = click(field,3,0);

  for(int i=0;i<num_rows;i++){
    for(int j=0;j<num_cols;j++){
      std::cout<<field[i][j]<<" ";
    }
    std::cout<<"\n";
  }

}

int main (){
  int bombs[3][2]= {{0,0},{3,3},{1,2}};
  int num_cols =5;
  int num_rows =5;
  mineSweeper(bombs,num_cols,num_rows);
  return 0;
}

File: ***Questions-Solutions***/Two-Dimensional-Arrays/Python/Q02MineSweeper.py
# Implement your function below.
def click(field, num_rows, num_cols, given_i, given_j):
    if field[given_i][given_j] == 1 or field[given_i][given_j] == -1:
        return field
    
    import Queue
    next_item = Queue.Queue()
    field[given_i][given_j] = -2
    next_item.put((given_i,given_j))
    while not next_item.empty():
        (current_i,current_j) = next_item.get()
        
        for i in range(current_i-1,current_i+2):
            for j in range(current_j-1,current_j+2):
                if 0<=i<num_rows and 0<=j<num_cols and field[i][j] == 0:
                    field[i][j] = -2
                    next_item.put((i,j))
       
                
    return field


# NOTE: Feel free to use the following function for testing.
# It converts a 2-dimensional array (a list of lists) into
# an easy-to-read string format.
def to_string(given_array):
    list_rows = []
    for row in given_array:
        list_rows.append(str(row))
    return '[' + ',\n '.join(list_rows) + ']'


# NOTE: The following input values will be used for testing your solution.
field1 = [[0, 0, 0, 0, 0],
          [0, 1, 1, 1, 0],
          [0, 1, -1, 1, 0]]

print(to_string(click(field1, 3, 5, 2, 2))) #should return:
# [[0, 0, 0, 0, 0],
#  [0, 1, 1, 1, 0],
#  [0, 1, -1, 1, 0]]

print(to_string(click(field1, 3, 5, 1, 4))) #should return:
# [[-2, -2, -2, -2, -2],
#  [-2, 1, 1, 1, -2],
#  [-2, 1, -1, 1, -2]]


field2 = [[-1, 1, 0, 0],
          [1, 1, 0, 0],
          [0, 0, 1, 1],
          [0, 0, 1, -1]]

print(to_string(click(field2, 4, 4, 0, 1))) #should return:
# [[-1, 1, 0, 0],
#  [1, 1, 0, 0],
#  [0, 0, 1, 1],
#  [0, 0, 1, -1]]

print(to_string(click(field2, 4, 4, 1, 3))) #should return:
# [[-1, 1, -2, -2],
#  [1, 1, -2, -2],
#  [-2, -2, 1, 1],
#  [-2, -2, 1, -1]]

